re_data_ui/src/partials/Graph.tsx

```diff
@@ -0,0 +1,328 @@
+import React, {
+  ReactElement, useContext, useEffect, useState,
+} from 'react';
+import { useSearchParams } from 'react-router-dom';
+import { NodeOptions, Options } from 'vis';
+import LineageGraph from '../components/LineageGraph';
+import ModelDetails from '../components/ModelDetails';
+import {
+  DbtNode, DbtSource, OverviewData, RedataOverviewContext,
+} from '../contexts/redataOverviewContext';
+import { supportedResTypes } from '../utils';
+
+export interface VisPointer {
+  x: number,
+  y: number
+}
+
+export interface VisNode extends NodeOptions {
+  id: string | number,
+  shape: string;
+}
+
+export interface VisEdge {
+  from: string | number,
+  to: string | number;
+  arrows: string;
+  color?: string;
+}
+
+export interface VisNetworkEventParams {
+  edges?: Array<string>,
+  nodes?: Array<string>,
+  event?: Record<string, unknown>,
+  pointer?: {
+    DOM: VisPointer
+    canvas: VisPointer
+  }
+}
+
+export interface IGraph {
+  nodes: Array<VisNode>;
+  edges: Array<VisEdge>;
+}
+
+type ResourceTypeColorsProps = {
+  [key: string]: string;
+}
+
+const resourceTypeColors: ResourceTypeColorsProps = {
+  source: 'hsl(97deg 66% 44%)',
+  model: 'hsl(190deg 100% 35%)',
+  seed: 'hsl(150deg 66% 44%)',
+};
+
+const generateGraph = (overview: OverviewData, modelName?: string | null) => {
+  const graph: IGraph = {
+    nodes: [],
+    edges: [],
+  };
+  if (!overview.graph) {
+    return graph;
+  }
+  const { dbtMapping } = overview;
+  const dbtNodes = overview.graph.nodes;
+  const dbtSources = overview.graph.sources;
+
+  const allNodes = { ...dbtNodes, ...dbtSources };
+
+  if (modelName) {
+    const { parent_map: parentNodes, child_map: childNodes } = overview.graph;
+    const modelTitle = dbtMapping[modelName];
+
+    // get all the parents and the child nodes of the model name;
+    const modelParentNodes = parentNodes[modelTitle] || [];
+    const modelChildNodes = childNodes[modelTitle] || [];
+
+    // draw the model node
+    // draw the parent nodes and connect it to the node using an edge
+    // draw the child nodes and connect it to the node using an edge
+
+    const details = allNodes[modelTitle];
+    const modelId = `${details.database}.${details.schema}.${details.name}`.toLowerCase();
+    graph.nodes.push({
+      id: modelId,
+      label: details.name,
+      shape: 'box',
+      color: resourceTypeColors[details.resource_type],
+    });
+
+    if (modelParentNodes.length) {
```

Comment:
is there a need for this check? `.forEach()` won't iterate if the array is empty.

Reply:
yeah, thanks